Fix #18003 regression by adding a check for borderless buttons

akiselev opened the pull request:
Fixes #18003

I created this silly regression with my second PR since I wasn't yet used to the appearance of some of the other buttons on the same page :facepalm:

Fix by checking for borderless buttons which covers the `<RunButtonWithTooltip>` in the `<ViewHeader>` component

**Steps to verify:**

1. Custom question > Sample Dataset > Orders
2. Do not click Visualize, just click "Hide editor" in top-right corner image


**Before (with regression):**
![image](https://user-images.githubusercontent.com/653604/134430264-2c4a35d6-f92f-4d43-bbf7-35f4a67e4fae.png)

**After:**
![image](https://user-images.githubusercontent.com/653604/134430292-026e99a9-197d-4c31-bcee-ad9cdc27550f.png)

Commit:
added check for borderless buttons when setting compact padding

Commit:
fix my own silly regression

pawit-metabase commented:
Can you fix this as well? Run button became bigger and off-centered in the native query page since #17953. (Chrome on macOS)

<img width="136" alt="Screen Shot 2564-09-23 at 14 22 41" src="https://user-images.githubusercontent.com/88995137/134468777-e35c08bd-3996-4625-b33c-23305a25f43b.png">

flamber commented:
@akiselev Everything is connected, so it can sometimes be difficult to change something in one place without breaking something somewhere else. It just takes a little while to get to know the code :wink: 
After you address what Ton pointed on in the Native editor, then you probably want to run visual testing:
https://github.com/metabase/metabase/blob/master/docs/developers-guide/visual-tests.md

pawit-metabase commented:
Re: visual tests

I've just checked and there's no screenshot that cover the run buttons right now, but it's definitely a good way to uncover this issue in the future. Perhaps you can find all the places the run button is used and add tests for them.

(Reach out to Reza or someone in the frontend team to get access to Percy if you don't have it already.)

gusaiani reviewed (approved):
👍

gusaiani commented:
Missed the comments that came before my approval.

Well, I suppose a visual regression test can be added here or in a separate PR, same for fixes of other instances of buttons.

Your call I suppose, Alex.

Commit:
added visual tests for run button regressions

akiselev commented:
Fixed the native query run button:
![image](https://user-images.githubusercontent.com/653604/134713901-c1177053-260b-44c7-ab12-89ccdce1150d.png)

Added two Percy visual tests to `frontend/test/metabase-visual/notebook/notebook.cy.spec.js`

akiselev commented:
@metabase-bot run visual tests

Commit:
added minWidth to percy snapshot

akiselev commented:
First pass at Percy didn't have a viewport wide enough for the full run button on the custom question page, trying the run again with minWidth.

Looks like there are a few other changes in the Percy run and I'm not qualified to review them. @pawit-metabase

akiselev commented:
@metabase-bot run visual tests

Commit:
shoudl have read the docs

akiselev commented:
@metabase-bot run visual tests

flamber reviewed (approved):
This looks good :+1: 
Let's see if @alxnddr can help, since he has played around in Percy the most.

alxnddr commented:
@akiselev the unrelated changes caused by changed dates:
<img width="279" alt="Screen Shot 2021-09-24 at 22 23 11" src="https://user-images.githubusercontent.com/14301985/134729284-a33fbd62-b1f1-4fa2-9de6-e8302155d18d.png">
Unfortunately, we do not have a way to deal with these false positives at the moment. Client-side dates can be mocked with Cypress, but these are coming from the BE. Feel free to approve them.

Did you solve the width issue?

Commit:
Sync VIEWPORT_WIDTH and percy snapshot width

akiselev commented:
I fixed the width issue but due to the way Percy works, the Cypress and Percy width config have to be the same. Percy serializes the DOM and rerenders it on their backend so if Cypress's width is set to 2500 but Percy's is 1920, some stuff renders correctly while others do not. In this case, the run button in the body of Custom question page is off centered.

Since Percy has a [max width of 2000px](https://docs.percy.io/docs/responsive-visual-testing#limits) I had to break it out into a separation description block  (in order to avoid changing the `visual tests > notebook > major UI elements` viewport width.

This last run should be good to go

akiselev commented:
@metabase-bot run visual tests

akiselev commented:
Setting the same width worked

Visual tests approved w/ nondeterministic date based changes and two new notebook tests.
